src/lib/pubkey/hss_lms/lms.cpp

```diff
@@ -0,0 +1,351 @@
+   auto lmots_params = LMOTS_Params::create_or_throw(lmots_type);
+
+   if(lms_params.hash_name() != lmots_params.hash_name()) {
+      throw Decoding_Error("No support for HSS-LMS instances with multiple hash functions.");
+   }
+
+   // Alg. 6. 2.e.
+   auto I = slicer.copy<LMS_Identifier>(LMS_IDENTIFIER_LEN);
+   // Alg. 6. 2.f.
+   auto lms_root = slicer.copy<LMS_Tree_Node>(lms_params.m());
+
+   return LMS_PublicKey(std::move(lms_params), std::move(lmots_params), std::move(I), std::move(lms_root));
+}
+
+std::vector<uint8_t> LMS_PublicKey::to_bytes() const {
+   std::vector<uint8_t> bytes(size(lms_params()));
+   BufferStuffer stuffer(bytes);
+
+   stuffer.append_be(lms_params().algorithm_type());
+   stuffer.append_be(lmots_params().algorithm_type());
+   stuffer.append(identifier());
+   stuffer.append(m_lms_root);
+   BOTAN_ASSERT_NOMSG(stuffer.full());
+
+   return bytes;
+}
+
+LMS_PublicKey::LMS_PublicKey(LMS_Params lms_params,
+                             LMOTS_Params lmots_params,
+                             LMS_Identifier I,
+                             LMS_Tree_Node lms_root) :
+      LMS_Instance(std::move(lms_params), std::move(lmots_params), std::move(I)), m_lms_root(std::move(lms_root)) {
+   BOTAN_ARG_CHECK(identifier().size() == LMS_IDENTIFIER_LEN, "Invalid LMS identifier");
+   BOTAN_ARG_CHECK(m_lms_root.size() == this->lms_params().m(), "Invalid LMS root");
+}
+
+size_t LMS_PublicKey::size(const LMS_Params& lms_params) {
+   return sizeof(LMS_Algorithm_Type) + sizeof(LMOTS_Algorithm_Type) + LMS_IDENTIFIER_LEN + lms_params.m();
+}
+
+LMS_Signature LMS_Signature::from_bytes_or_throw(BufferSlicer& slicer) {
+   size_t total_remaining_bytes = slicer.remaining();
+   // Alg. 6a 1. (next 4 bytes are checked in LMOTS_Signature::from_bytes_or_throw)
+   if(total_remaining_bytes < sizeof(LMS_Tree_Node_Idx)) {
+      throw Decoding_Error("To few signature bytes while parsing LMS signature.");
+   }
+   // Alg. 6a 2.a.
+   auto q = slicer.copy_be<LMS_Tree_Node_Idx>();
+
+   // Alg. 6a 2.b.-e.
+   auto lmots_sig = LMOTS_Signature::from_bytes_or_throw(slicer);
+   LMOTS_Params lmots_params = LMOTS_Params::create_or_throw(lmots_sig.algorithm_type());
+
+   if(slicer.remaining() < sizeof(LMS_Algorithm_Type)) {
+      throw Decoding_Error("To few signature bytes while parsing LMS signature.");
+   }
+   // Alg. 6a 2.f.
+   auto lms_type = slicer.copy_be<LMS_Algorithm_Type>();
+   // Alg. 6a 2.h.
+   LMS_Params lms_params = LMS_Params::create_or_throw(lms_type);
+   // Alg. 6a 2.i. (signature is not exactly [...] bytes long)
+   if(total_remaining_bytes < size(lms_params, lmots_params)) {
+      throw Decoding_Error("To few signature bytes while parsing LMS signature.");
+   }
+
+   // Alg. 6a 2.j.
+   auto auth_path = slicer.take<LMS_AuthenticationPath>(lms_params.m() * lms_params.h());
+
+   return LMS_Signature(q, std::move(lmots_sig), lms_type, LMS_AuthenticationPath(auth_path));
```

Comment:
```suggestion
   return LMS_Signature(q, std::move(lmots_sig), lms_type, std::move(auth_path));
```

... looks like some refactoring left over to me.

Reply:
This is indeed a copy since we decided to copy the auth path into the signature object instead of storing only spans. Better would be to directly call `slicer.copy` instead of  `slicer.take`, though.